components/collator/src/comparison.rs

```diff
@@ -1543,4 +1644,660 @@ impl CollatorBorrowed<'_> {
+                        } else {
+                            SEC_COMMON[WEIGHT_HIGH] - remainder
+                        };
+                        secondaries.append_byte(b);
+                        common_secondaries -= remainder;
+                        // common_secondaries is now a multiple of SEC_COMMON[WEIGHT_MAX_COUNT]
+                        while common_secondaries > 0 {
+                            // same as >= SEC_COMMON[WEIGHT_MAX_COUNT]
+                            secondaries.append_byte(SEC_COMMON[WEIGHT_MIDDLE]);
+                            common_secondaries -= SEC_COMMON[WEIGHT_MAX_COUNT];
+                        }
+                        // commonSecondaries == 0
+                    }
+                    if 0 < p && p <= MERGE_SEPARATOR_PRIMARY {
+                        // The backwards secondary level compares secondary weights backwards
+                        // within segments separated by the merge separator (U+FFFE).
+                        let secs = secondaries.as_mut();
+                        let last = secs.len() - 1;
+                        if sec_segment_start < last {
+                            let mut q = sec_segment_start;
+                            let mut r = last;
+
+                            // these indices start at valid values and we stop when they cross
+                            #[allow(clippy::indexing_slicing)]
+                            while q < r {
+                                let b = secs[q];
+                                secs[q] = secs[r];
+                                q += 1;
+                                secs[r] = b;
+                                r -= 1;
+                            }
+                        }
+                        let b = if p == NO_CE_PRIMARY {
+                            LEVEL_SEPARATOR_BYTE
+                        } else {
+                            MERGE_SEPARATOR_BYTE
+                        };
+                        secondaries.append_byte(b);
+                        prev_secondary = 0;
+                        sec_segment_start = secondaries.len();
+                    } else {
+                        secondaries.append_reverse_weight_16(s);
+                        prev_secondary = s;
+                    }
+                }
+            }
+
+            if levels & CASE_LEVEL_FLAG != 0 {
+                if self.options.strength() == Strength::Primary && p == 0
+                    || non_primary.bits() <= 0xffff
+                {
+                    // Primary+caseLevel: Ignore case level weights of primary ignorables.
+                    // Otherwise: Ignore case level weights of secondary ignorables.  For
+                    // details see the comments in the CollationCompare class.
+                } else {
+                    // case bits & tertiary lead byte
+                    let mut c = ((non_primary.bits() >> 8) & 0xff) as u8;
+                    assert_ne!(c & 0xc0, 0xc0);
```

Comment:
Please downgrade the assertions to debug assertions. It seems to me that the assertions added in this PR are not needed for memory-safety, and we generally want to avoid panics even if the alternative is bogus output.